service/history/queues/iterator.go

```diff
@@ -0,0 +1,145 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package queues
+
+import (
+	"fmt"
+
+	"go.temporal.io/server/common/collection"
+	"go.temporal.io/server/service/history/tasks"
+)
+
+type (
+	Iterator interface {
+		collection.Iterator[tasks.Task]
+
+		Range() tasks.Range
+		CanSplit(tasks.Key) bool
+		Split(key tasks.Key) (left Iterator, right Iterator)
+		CanMerge(Iterator) bool
+		Merge(Iterator) Iterator
+		Remaining() Iterator
+	}
+
+	paginationFnProvider func(tasks.Range) collection.PaginationFn[tasks.Task]
+
+	IteratorImpl struct {
+		paginationFnProvider paginationFnProvider
+		remainingRange       tasks.Range
+
+		pagingIterator collection.Iterator[tasks.Task]
+	}
+)
+
+func NewIterator(
+	paginationFnProvider paginationFnProvider,
+	r tasks.Range,
```

Comment:
`r tasks.Range,` -> `range tasks.Range,`

Reply:
`range` is reserved keyword.